apps/checkout-app/pages/api/pay.ts

```diff
@@ -65,9 +73,78 @@ async function handler(req: NextApiRequest, res: NextApiResponse) {
 
   let response: PayRequestResponse;
 
+  if (order.privateMetafield) {
+    const payment: OrderPaymentMetafield = JSON.parse(order.privateMetafield);
+
+    if (payment.provider === body.provider && payment.session) {
+      if (payment.provider === "mollie") {
+        const session = await verifyMollieSession(payment.session);
+
+        if (session.status === MollieOrderStatus.created && session.url) {
+          response = {
+            ok: true,
+            provider: payment.provider,
+            orderId: order.id,
+            data: {
+              paymentUrl: session.url,
+            },
+          };
+
+          return res.status(200).json(response);
+        } else if (
+          [
+            MollieOrderStatus.authorized,
+            MollieOrderStatus.completed,
+            MollieOrderStatus.paid,
+            MollieOrderStatus.pending,
+            MollieOrderStatus.shipping,
+          ].includes(session.status)
+        ) {
+          response = {
+            ok: false,
+            provider: payment.provider,
+            orderId: order.id,
+            errors: ["ALREADY_PAID"],
+          };
+
+          return res.status(200).json(response);
+        }
+      } else if (payment.provider === "adyen") {
+        const session = await verifyAdyenSession(payment.session);
+        const StatusEnum = AdyenTypes.checkout.PaymentLinkResource.StatusEnum;
+
+        if (session.status === StatusEnum.Active) {
+          response = {
+            ok: true,
+            provider: payment.provider,
+            orderId: order.id,
+            data: {
+              paymentUrl: session.url,
+            },
+          };
+
+          return res.status(200).json(response);
+        } else if (
+          [StatusEnum.Completed, StatusEnum.PaymentPending].includes(
+            session.status
+          )
+        ) {
+          response = {
+            ok: false,
+            provider: payment.provider,
+            orderId: order.id,
+            errors: ["ALREADY_PAID"],
+          };
+
+          return res.status(200).json(response);
+        }
+      }
+    }
+  }
+
   if (body.provider === "mollie") {
```

Comment:
Ok so this function has gotten srsly long, maybe we can do something to split it into smaller ones? I think maybe the contents of `if (body.provider === "mollie") {` and similarly `} else if (body.provider === "adyen") {` could be moved to separate functions, as well as parts for earlier adyen and mollie sections

Reply:
I've split up added logic into new fn